domain/credential/state/state.go

```diff
@@ -325,60 +397,120 @@ WHERE  cloud.name = $dbCloudName.name
+// invalid.
+// The following errors can be expected:
+// - [credentialerrors.NotFound] when no credential is found for the
+// given uuid.
+func (st *State) InvalidateCloudCredential(ctx context.Context, uuid corecredential.UUID, reason string) error {
 	db, err := st.DB()
 	if err != nil {
 		return errors.Capture(err)
 	}
 
-	q := `
-UPDATE cloud_credential
-SET    invalid = true, invalid_reason = $M.invalid_reason
-FROM cloud
-WHERE  cloud_credential.name = $M.credential_name
-AND    cloud_credential.owner_uuid = (
-    SELECT uuid
-    FROM user
-    WHERE user.name = $M.owner
-	AND user.removed = false
-)
-AND    cloud_credential.cloud_uuid = (
-    SELECT uuid FROM cloud
-    WHERE name = $M.cloud_name
-)`
-	stmt, err := st.Prepare(q, sqlair.M{})
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return st.invalidateCloudCredential(ctx, tx, uuid, reason)
+	})
+}
+
+// InvalidateModelCloudCredential marks the cloud credential that is in use by
+// the given model as invalid. This will affect not just the model used to find
+// the cloud credential but all models that are using the same cloud credential
+// as the model provided.
+//
+// This function will only work with models that are active.
+// The following errors can be expected:
+// - [modelerrors.NotFound] if the no model exists for the provided uuid.
+func (st *State) InvalidateModelCloudCredential(
+	ctx context.Context,
+	uuid coremodel.UUID,
+	reason string,
+) error {
+	db, err := st.DB()
 	if err != nil {
 		return errors.Capture(err)
 	}
 
-	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		var outcome sqlair.Outcome
-		terms := sqlair.M{
-			"credential_name": key.Name,
-			"cloud_name":      key.Cloud,
-			"owner":           key.Owner.Name(),
-		}
-		terms["invalid_reason"] = reason
-		err = tx.Query(ctx, stmt, terms).Get(&outcome)
-		if err != nil {
-			return errors.Capture(err)
-		}
-		n, err := outcome.Result().RowsAffected()
-		if err != nil {
-			return errors.Capture(err)
+	modelUUID := modelUUID{UUID: uuid.String()}
+	modelCredentialUUID := modelCredentialUUID{}
+	stmt, err := st.Prepare(`
+SELECT &modelCredentialUUID.*
+FROM v_model
+WHERE uuid = $modelUUID.uuid`,
+		modelUUID, modelCredentialUUID)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx, stmt, modelUUID).Get(&modelCredentialUUID)
+		if errors.Is(err, sql.ErrNoRows) {
+			return errors.Errorf(
+				"model %q does not exist", uuid,
+			).Add(modelerrors.NotFound)
+		} else if err != nil {
+			return errors.Errorf(
+				"getting cloud credential uuid for model %q: %w",
+				uuid, err,
+			)
 		}
-		if n < 1 {
-			return errors.Errorf("credential %q for cloud %q owned by %q %w", key.Name, key.Cloud, key.Owner, coreerrors.NotFound)
+
+		// The model doesn't have a credential set so there is nothing to do.
+		if !modelCredentialUUID.UUID.Valid {
+			return nil
 		}
```

Comment:
Return a `ModelCredentialNotSet` error here and the caller can decide how to handle that. This layer should not make that determination.

Reply:
Done & Done